File: backend/agents/writer.py
from google import genai
from google.genai import types
import os
import json
from dotenv import load_dotenv
from pydantic import BaseModel, Field
from typing import List, Optional

load_dotenv()

class Scene(BaseModel):
    sequenceNumber: int
    script: str
    visual_prompt: str
    rationale: str

class WriterOutput(BaseModel):
    scenes: List[Scene]
    thought_signature: Optional[str] = None

class WriterAgent:
    def __init__(self):
        self.client = genai.Client(api_key=os.getenv("GOOGLE_API_KEY"))
        self.model_name = "gemini-3-pro-preview" # Using Gemini 3 Pro Preview
        # TODO: Monitor for stability updates.

    def generate_script(self, topic: str, genre: str, tone: str, previous_thought_signature: str = None) -> WriterOutput:
        
        system_instruction = f"""
        You are a professional screenwriter.
        Your task is to write a short video script based on the given topic, genre, and tone.
        
        Output must be a JSON object with a list of 'scenes'.
        Each scene must have:
        - sequenceNumber: integer
        - script: dialogue or narration
        - visual_prompt: detailed description for an image generator
        - rationale: why you wrote this script and visual choice (your director's note)
        
        Genre: {genre}
        Tone: {tone}
        """

        # Thinking config (if supported by the model/SDK version)
        # Currently simulating or using standard generation config
        config = types.GenerateContentConfig(
            temperature=0.7,
            response_mime_type="application/json",
            response_schema=WriterOutput,
            # thinking_level="high", # Hypothetical parameter from spec
        )

        prompt = f"Topic: {topic}"
        
        # If we had a previous thought signature, we would inject it here or in the context
        if previous_thought_signature:
            prompt += f"\n\nPrevious Context/Thought: {previous_thought_signature}"

        try:
            response = self.client.models.generate_content(
                model=self.model_name,
                contents=prompt,
                config=config
            )
            
            # Extracting the parsed object
            # The SDK might return a parsed object directly if response_schema is set
            if response.parsed:
                return response.parsed
            else:
                # Fallback parsing
                return WriterOutput.model_validate_json(response.text)

        except Exception as e:
            print(f"Error in WriterAgent: {e}")
            raise e

if __name__ == "__main__":
    # Simple test
    agent = WriterAgent()
    try:
        result = agent.generate_script("A cat exploring a futuristic city", "Sci-Fi", "Wonder")
        print(json.dumps(result.model_dump(), indent=2))
    except Exception as e:
        print(e)

File: src/agents/writer.ts
import { GoogleGenerativeAI, SchemaType, Schema } from "@google/generative-ai";

// --- Interfaces ---

export interface ProductionBible {
    genre: string;
    tone: string;
    visualStyle: string;
    characters: Record<string, {
        name: string;
        description: string;
        referenceImageUrls: string[];
    }>;
}

export interface WriterInput {
    topic: string;
    productionBible: ProductionBible;
    previousThoughtSignature?: string;
}

export interface Scene {
    sceneId: string;
    sequenceNumber: number;
    script: string;
    visualPrompt: string;
    rationale: string; // Explanation for why this scene is engaging
}

export interface WriterOutput {
    scenes: Scene[];
    globalRationale: string; // Overall explanation of the script structure
    thoughtSignature: string; // Encrypted token for context maintenance
}

// --- System Prompt ---

const WRITER_SYSTEM_PROMPT = `
You are a world-class Short-Form Video Writer Agent specializing in dopamine-inducing storytelling.
Your goal is to create a script that captures the viewer's attention within the first 3 seconds (The Hook) and maintains high engagement throughout.

**Role & Responsibilities:**
1.  **Dopamine Architect:** Design scenes that trigger curiosity, excitement, or shock.
2.  **Visual Storyteller:** Write visual prompts that are vivid, cinematic, and specific (lighting, camera angles, action).
3.  **Rationale Provider:** You MUST explain your creative decisions. For every scene, provide a 'rationale' explaining WHY it works psychologically or narratively.
4.  **Context Keeper:** You will receive a 'previousThoughtSignature' if this is a continuation. Use it to maintain consistency.

**Output Format:**
You must output strictly valid JSON matching the defined schema.
The output must include an array of scenes, a global rationale, and a thought signature.

**Constraints:**
- The first scene MUST be a strong hook.
- Keep dialogue concise and impactful.
- Ensure visual prompts are detailed enough for an AI Artist agent.
`;

// --- Implementation ---

// const genAI = new GoogleGenerativeAI(process.env.NEXT_PUBLIC_GEMINI_API_KEY || "");

const schema: Schema = {
    type: SchemaType.OBJECT,
    properties: {
        scenes: {
            type: SchemaType.ARRAY,
            items: {
                type: SchemaType.OBJECT,
                properties: {
                    sceneId: { type: SchemaType.STRING },
                    sequenceNumber: { type: SchemaType.NUMBER },
                    script: { type: SchemaType.STRING },
                    visualPrompt: { type: SchemaType.STRING },
                    rationale: { type: SchemaType.STRING },
                },
                required: ["sceneId", "sequenceNumber", "script", "visualPrompt", "rationale"],
            },
        },
        globalRationale: { type: SchemaType.STRING },
        thoughtSignature: { type: SchemaType.STRING },
    },
    required: ["scenes", "globalRationale", "thoughtSignature"],
};

export async function generateScript(input: WriterInput): Promise<WriterOutput> {
    const genAI = new GoogleGenerativeAI(process.env.NEXT_PUBLIC_GEMINI_API_KEY || "");
    const model = genAI.getGenerativeModel({
        model: "gemini-3-pro-preview",
        generationConfig: {
            responseMimeType: "application/json",
            responseSchema: schema,
            // @ts-ignore - thinking_level is a preview feature not yet in official types
            // thinking_level: "high",
        },
        systemInstruction: WRITER_SYSTEM_PROMPT,
    });

    const prompt = `
    Topic: ${input.topic}
    
    Production Bible:
    ${JSON.stringify(input.productionBible, null, 2)}
    
    ${input.previousThoughtSignature ? `Previous Thought Signature: ${input.previousThoughtSignature}` : ""}
    
    Generate a short-form video script based on the above input.
  `;

    try {
        const result = await model.generateContent(prompt);
        const response = result.response;
        const text = response.text();

        // Parse JSON output
        const output = JSON.parse(text) as WriterOutput;
        return output;
    } catch (error) {
        console.error("Error generating script:", error);
        throw new Error("Failed to generate script from Writer Agent.");
    }
}
